xdsl/parser/attribute_parser.py

```diff
@@ -504,20 +519,46 @@ def parse_shape_delimiter(self) -> None:
         # Move the lexer to the position after 'x'.
         self._resume_from(self._current_token.span.start + 1)
 
+    def parse_dimension_list(self) -> list[int]:
+        """
+        Parse a dimension list with the following format:
+          dimension-list ::= (dimension `x`)* dimension
+        each dimension is also required to be non-negative.
+        """
+        dims: list[int] = []
+        accepted_token_kinds = (MLIRTokenKind.INTEGER_LIT, MLIRTokenKind.QUESTION)
+
+        # empty case
+        if self._current_token.kind not in accepted_token_kinds:
+            return []
+
+        # parse first number
+        dim = self.parse_shape_dimension()
+        dims.append(dim)
+
+        # Two approaches possible here, could look for `x` `num` and only parse
+        # when both are found (requires peek() functionality), or could parse
+        # the `x` always, and then `undo` if we don't see a number following
+        # which is this approach
```

Comment:
I like this approach, let's remove this commend and add one below to make it clear whey we're resuming from a different position